pyanaconda/modules/subscription/subscription.py

```diff
@@ -26,3 +38,183 @@ class SubscriptionService(KickstartService):
+        role = self._process_role(data)
+        sla = self._process_sla(data)
+        usage = self._process_usage(data)
+
+        system_purpose_data = SystemPurposeData()
+        system_purpose_data.role = role
+        system_purpose_data.sla = sla
+        system_purpose_data.usage = usage
+
+        if data.syspurpose.addons:
+            # As we do not have a list of valid addons available, we just use what was provided
+            # by the user in kickstart verbatim.
+            system_purpose_data.addons = data.syspurpose.addons
+
+        self.set_system_purpose_data(system_purpose_data)
+
+    def setup_kickstart(self, data):
+        """Return the kickstart string."""
+        log.debug("Generating kickstart data...")
+
+        # system purpose
+        data.syspurpose.role = self.system_purpose_data.role
+        data.syspurpose.sla = self.system_purpose_data.sla
+        data.syspurpose.usage = self.system_purpose_data.usage
+        data.syspurpose.addons = self.system_purpose_data.addons
+
+        return str(data)
+
+    # system purpose configuration
+
+    def _load_valid_system_purpose_values(self):
+        """Load lists of valid roles, SLAs and usage types.
+
+        About role/sla/validity:
+        - an older installation image might have older list of valid fields,
+          missing fields that have become valid after the image has been released
+        - fields that have been valid in the past might be dropped in the future
+        - there is no list of valid addons
+
+        Due to this we need to take into account that the listing might not always be
+        comprehensive and that we need to allow what might on a first glance look like
+        invalid values to be written to the target system.
+        """
+        roles, slas, usage_types = system_purpose.get_valid_fields()
+        self._valid_roles = roles
+        self._valid_slas = slas
+        self._valid_usage_types = usage_types
+
+    def _process_role(self, data):
+        if data.syspurpose.role:
+            role_match = system_purpose.match_field(data.syspurpose.role, self.valid_roles)
+        else:
+            role_match = None
+
+        if role_match:
+            log.info("role value %s from kickstart matched to known valid field %s",
+                     data.syspurpose.role,
+                     role_match)
+            return role_match
+        elif data.syspurpose.role:
+            log.info("using custom role value from kickstart: %s", data.syspurpose.role)
+            return data.syspurpose.role
+        else:
+            return ""
+
+    def _process_sla(self, data):
+        if data.syspurpose.sla:
+            sla_match = system_purpose.match_field(data.syspurpose.sla, self.valid_slas)
+        else:
+            sla_match = None
+
+        if sla_match:
+            log.info("SLA value %s from kickstart matched to known valid field %s",
+                     data.syspurpose.sla,
+                     sla_match)
```

Comment:
Write this on two lines, I don't think there is a reason to add each parameter on a separate line in this case.

Reply:
Code has been refactored and moved to the system_purpose library.